### PR TITLE
New cmdlet with working unit test.

### PR DESCRIPTION
**Pull request by zippy1981**
Implements #1311

So it might make sense to add casts or something to make this more useful.

Changes proposed in this pull request:
 - Get-SqlConnectionStringBuilder
 - 
 - 

How to test this code: 
- [ ] Throw connection strings at it
- [ ] Modify connection strings with the resulting objects
- [] Connect to databases with those connection strings

Has been tested on minimum requirements:
- [ ]  Powershell 3
- [ ]  Windows 7
- [ ]  SQL Server 2000

Has been tested on maximum requirements:
- [ ]  SQL Server vNext
- [X]  Windows 10
- [ ]  Azure Database

Tests for tester:
- [ ] Working/useful help content, including link to command on dbatools web site
- [ ] All examples work as advertised
- [ ] Does not contain template content
- [ ] Does not contain excessive/unnecessary amounts of comments
- [ ] Works remotely
- [ ] Works locally
- [ ] Works on lower versions or throws error specifying version not supported
- [ ] Works with named instances
- [ ] Works with clustered instances
- [ ] Handles offline/read only databases
- [ ] Supports multiple servers (at the command line or piped from Get-SqlRegisteredServerName)
- [ ] No un-handled errors which stop the command working with multiple servers

### CODE REVIEW

File: functions/Get-DbaSqlConnectionStringBuilder.ps1 [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+Returns an empty ConnectionStringBuilder
+
+.EXAMPLE
+"Data Source=localhost,1433;Initial Catalog=AlwaysEncryptedSample;UID=sa;PWD=alwaysB3Encrypt1ng;Application Name=Always Encrypted Sample MVC App;Column Encryption Setting=enabled" | Get-DbaSqlConnectionStringBuilder 
+
+Returns a connection string builder that can be used to connect to the local sql server instance on the default port.
+
+#>
+	[Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSUseShouldProcessForStateChangingFunctions", "")]
```
Comment (FriedrichWeinmann) [inline functions/Get-DbaSqlConnectionStringBuilder.ps1]: It is not necessary to exclude this test, es the function does not violate it.

File: tests/Get-DbaSqlConnectionStringBuilder.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,14 @@
+Describe "Get-DbaSqlConnectionStringBuilder Tests" {
```
Comment (FriedrichWeinmann) [inline tests/Get-DbaSqlConnectionStringBuilder.Tests.ps1]: You'll want to update this line to this:
`Describe "Get-DbaSqlConnectionStringBuilder Unit Tests" -Tag 'Unittests' {`

File: functions/Get-DbaSqlConnectionStringBuilder.ps1 [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	[CmdletBinding()]
+	Param (
+		[Parameter(Mandatory = $false, ValueFromPipeline = $true)]
+        [string[]]$Connectiontring = $null
```
Comment (FriedrichWeinmann) [inline functions/Get-DbaSqlConnectionStringBuilder.ps1]: Do I detect a typo here?

File: functions/Get-DbaSqlConnectionStringBuilder.ps1 [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	)
+    process {
+        New-Object Data.SqlClient.SqlConnectionStringBuilder $Connectiontring
```
Comment (FriedrichWeinmann) [inline functions/Get-DbaSqlConnectionStringBuilder.ps1]: This cannot handle multiple connectionstring inputs that were _not_ passed by pipeline.
You'll need to wrap this into a foreach-loop, calling new-Object for each string in `$ConnectionString`

File: tests/Get-DbaSqlConnectionStringBuilder.Tests.ps1 (hunk 2 above)
Comment (zippy1981) [inline tests/Get-DbaSqlConnectionStringBuilder.Tests.ps1]: done

File: functions/Get-DbaSqlConnectionStringBuilder.ps1 (hunk 3 above)
Comment (zippy1981) [inline functions/Get-DbaSqlConnectionStringBuilder.ps1]: Yes

File: functions/Get-DbaSqlConnectionStringBuilder.ps1 (hunk 1 above)
Comment (zippy1981) [inline functions/Get-DbaSqlConnectionStringBuilder.ps1]: Fixed

**Comment by zippy1981** [comment]
@FriedrichWeinmann I'll work on adding more paramaters, and maybe making the output not be Key/Value pairs. I also got to get the foreach working right.

File: functions/New-DbaSqlConnectionStringBuilder.ps1 [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+.LINK
+https://dbatools.io/Get-DbaSqlConnectionStringBuilder
+
+.EXAMPLE
+Get-DbaSqlConnectionStringBuilder
+
+Returns an empty ConnectionStringBuilder
+
+.EXAMPLE
+"Data Source=localhost,1433;Initial Catalog=AlwaysEncryptedSample;UID=sa;PWD=alwaysB3Encrypt1ng;Application Name=Always Encrypted Sample MVC App;Column Encryption Setting=enabled" | Get-DbaSqlConnectionStringBuilder
```
Comment (FriedrichWeinmann) [inline functions/New-DbaSqlConnectionStringBuilder.ps1]: Both examples should be using the correct command name after renaming it

File: functions/New-DbaSqlConnectionStringBuilder.ps1 [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+Returns a connection string builder that can be used to connect to the local sql server instance on the default port.
+
+#>
+	[CmdletBinding()]
+	Param (
+		[Parameter(Mandatory = $false, ValueFromPipeline = $true)]
+        [string[]]$ConnectionString = $null
+	)
+    process {
+		New-Object Data.SqlClient.SqlConnectionStringBuilder $ConnectionString
```
Comment (FriedrichWeinmann) [inline functions/New-DbaSqlConnectionStringBuilder.ps1]: You can do the loop like this:
```PowerShell
foreach ($string in $ConnectionString) {
    New-Object Data.SqlClient.SqlConnectionStringBuilder $string
}
```

**Comment by zippy1981** [comment]
@FriedrichWeinmann I added the loop. I can't get it working when I pass an array to it can you try? I began adding paramaters and a few more test. I have to finish.

This change brought up a greater issue. @potatoqualitee @SQLDBAWithABeard Do you think that dbatools should try to use the connection string "dbatools Powershell Module" when making a connection string? Obviously don't override one actually set by the user, just its good to have a more specific default.

**Comment by potatoqualitee** [comment]
hey @zippy1981 - naw, I'd go with something even more generic - perhaps even take their username and be like "app by $env:domain\$env:username" (or whatever the format is. So they can see that it exists and remove or modify it.

**Comment by potatoqualitee** [comment]
I haven't gotten a chance to look at this yet, but I wanted to suggest (if i hadn't before) that you could take a look at Connect-DbaSqlServer - that has a ton of good stuff to help build the connection string. And if you don't use a .Connect(), you could just rip the $server.connectioncontext.connectionstring that it auto builds.

**Comment by potatoqualitee** [comment]
oops,i was way off LOL. great work! i fixed the loop and merged 👍

**Pull request merged**